sdk/include/opentelemetry/sdk/metrics/aggregator/histogram_aggregator.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include "opentelemetry/version.h"
+#include "opentelemetry/metrics/instrument.h"
+#include "opentelemetry/sdk/metrics/aggregator/aggregator.h"
+
+#include <variant>
+#include <vector>
+#include <mutex>
+#include <memory>
+
+namespace metrics_api = opentelemetry::metrics;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+template<class T>
+class HistogramAggregator final : public Aggregator<T>
+{
+    
+public:
+    
+    /**
+     * Constructor for the histogram aggregator.  A sorted vector of boundaries is expected and boundaries are
+     * doubles regardless of the aggregator's templated data type.
+     *
+     * Sum is stored in values_[0]
+     * Count is stored in position_[1]
+     */
+    HistogramAggregator(metrics_api::BoundInstrumentKind kind, std::vector<double> boundaries)
+    {
+        this->kind_ = kind;
+        boundaries_   = boundaries;
+        this->values_     = std::vector<T>(2, 0);
+        this->checkpoint_ = std::vector<T>(2, 0);
+        bucketCounts_ = std::vector<int>(boundaries_.size() + 1, 0);
+    }
+    
+    /**
+     * Recieves a captured value from the instrument and inserts it into the current histogram counts.
+     * Depending on the use case, linear search or binary search based implementations may be preferred.
+     * In a uniformly distributed dataset, linear search outperforms binary search until 512 buckets.
+     *
+     * @param val, the raw value used in aggregation
+     * @return none
+     */
+    void update(T val) override
+    {
+        
+        int bucketID = boundaries_.size();
+        for (int i = 0; i < boundaries_.size(); i++)
+        {
+            if (val < boundaries_[i]) // concurrent read is thread-safe
+            {
+                bucketID = i;
+                break;
+            }
+        }
+        
+        // Alternate implementation with binary search
```

Comment:
It'll be helpful to put a comment on the design decision - e.g. why would we choose linear search vs. binary search.

Reply:
Included some background on the topic above the update function -- let me know if this is not enough and I'll gladly add a more thorough explanation.